src/canopy/llm/openai.py

```diff
@@ -164,17 +159,24 @@ def enforced_function_call(self,
 
         messages = [m.dict() for m in messages]
 
-        chat_completion = openai.ChatCompletion.create(
+        # this enforces the model to call the function
+        function_call = completion_create_params.ChatCompletionFunctionCallOptionParam(
+            name=function.name)
```

Comment:
From my understanding of the [documentation](https://github.com/openai/openai-python/tree/main#nested-params), you're not supposed to instantiate these objects and pass them to the API. I don't see them instantiating typed objects anywhere in their examples.  
You pass a regular `dict`, and they enforce the typing behind the scene (raising a runtime error if the type is wrong).

Reply:
changed